src/OpenIddict.EntityFramework/Stores/OpenIddictApplicationStore.cs

```diff
@@ -651,7 +651,19 @@ public virtual ValueTask<ImmutableArray<string>> GetRedirectUrisAsync([NotNull]
         /// whose result returns the instantiated application, that can be persisted in the database.
         /// </returns>
         public virtual ValueTask<TApplication> InstantiateAsync(CancellationToken cancellationToken)
-            => new ValueTask<TApplication>(new TApplication());
+        {
+            try
+            {
+                return new ValueTask<TApplication>(Activator.CreateInstance<TApplication>());
+            }
+            catch (MissingMethodException exception)
+            {
+                throw new OpenIddictException(OpenIddictConstants.Exceptions.InstantiationError, new StringBuilder()
```

Comment:
It would be better to throw the exception "asynchronously". E.g:

```csharp
try
{
    return new ValueTask<TApplication>(Activator.CreateInstance<TApplication>());
}

catch (MissingMemberException exception)
{
    return new ValueTask<TApplication>(Task.FromException<TApplication>(
        new OpenIddictException(OpenIddictConstants.Exceptions.InstantiationError, new StringBuilder()
            .AppendLine("An error occurred while trying to create a new application instance.")
            .Append("Make sure that the application entity has a public parameterless constructor or create ")
            .Append("a custom store that overrides the 'InstantiateAsync()' method to use a custom factory.")
            .ToString(), exception)));
}

catch (MemberAccessException exception)
{
    return new ValueTask<TApplication>(Task.FromException<TApplication>(
        new OpenIddictException(OpenIddictConstants.Exceptions.InstantiationError, new StringBuilder()
            .AppendLine("An error occurred while trying to create a new application instance.")
            .Append("Make sure that the application entity is not an abstract class or create a ")
            .Append("custom store that overrides the 'InstantiateAsync()' method to use a custom factory.")
            .ToString(), exception)));
}
```

Reply:
BTW, I'm not sure it's worth using `OpenIddictException` here. `InvalidOperationException` might be better.